guidance/library/_json.py

```diff
@@ -359,16 +360,24 @@ def _gen_json_object(
     required: Sequence[str],
     definitions: Mapping[str, Callable[[], GrammarFunction]],
 ):
-    if any(k not in properties for k in required):
-        raise ValueError(f"Required properties not in properties: {set(required) - set(properties)}")
-
-    grammars = tuple(f'"{name}":' + _gen_json(json_schema=schema, definitions=definitions) for name, schema in properties.items())
-    required_items = tuple(name in required for name in properties)
+    # "required" keys will be validated against "properties" if they're present, otherwise against "additionalProperties".
+    # If "additionalProperties" is False, then required keys must be in "properties".
+    if any(k not in properties for k in required) and additional_properties is False:
+        raise ValueError(
+            f"Required properties not in properties but additionalProperties is False."
+            f" Missing required properties: {set(required) - set(properties)}"
+        )
+    items = list(chain(
+        # First iterate over the properties in order
+        properties.items(),
+        # If there are any keys in required that weren't specified by properties, add them in order at the end,
+        # where we will validate against the additional_properties schema
+        ((key, additional_properties) for key in required if key not in properties),
+    ))
+    grammars = tuple(f'"{name}":' + _gen_json(json_schema=schema, definitions=definitions) for name, schema in items)
+    required_items = tuple(name in required for name, _ in items)
 
     if additional_properties is not False:
-        if additional_properties is True:
```

Comment:
Not a functional change -- this line is just no longer needed since `_gen_json` can take bool schemas just fine